### DISCUSSION TITLE
Encoding of Rollups inputs

### DISCUSSION THREAD

**Discussion by augustoteixeira** [cartesi/cips#15] [General]
It would be beneficial to have a basic encoding of inputs arriving to the Cartesi Machine. One first reason for this is that it would allow us to differentiate between direct inputs (which are sent straight to the application) and aggregated inputs (which should be split before).

Here is an initial draft:

https://github.com/cartesi/cips/blob/aggregation/cips/cip-4.md

@felipeargento, please let me know what you think.

**Comment by guidanoli** [comment]
Not to be pedantic, but I do think it's important to point out in the CIP what interval of values each encoding supports.
For example, for the value `0xFF`, one could technically encode it with:

* 1 byte (`01111111`)
* 2 bytes (`10000000.01111111`)
* 3 bytes (`11000000.00000000.01111111`)
* 4 bytes (`11100000.00000000.00000000.01111111`)

So, as in the [UTF-8 specification](https://datatracker.ietf.org/doc/html/rfc3629#section-3) referenced in the CIP, we should impose that:

* 1 byte encoding only supports values from `0x00` to `0x7F`
* 2 bytes encoding only supports values from `0x80` to `0x3FFFF`
* 3 bytes encoding only supports values from `0x40000` to `0x1FFFFF`
* 4 bytes encoding only supports values from `0x200000` to `0xFFFFFFF`

This guarantees that the cost of data availability is minimized, and provides bijectivity to the conversion between the protocol number and its encoded representation.

**Reply by augustoteixeira** [reply]
This is a really good point! Could you add this into a PR?

**Reply by guidanoli** [reply]
Sure!

**Comment by felipeargento** [comment]
Cool!
How do you see aggregation of inputs from different protocols? Would it be a protocol that indicates that the inner inputs have different protocols?

Do you see protocols being developed for specific types of inputs? Meaning that we could design a protocol that achieves maximal compression for financial transactions, for instance.

**Reply by augustoteixeira** [reply]
I was thinking that we can implement what you are considering using "two levels of protocols". Something like:

- This is an input of type "Aggregated inputs" (this one even has a CIP already);
- The Transaction Protocol used in the aggregated inputs could allow for several types of inputs.

**Comment by guidanoli** [comment]
I did some research on [UTF-8](https://en.wikipedia.org/wiki/UTF-8) and collected the following points in favor of using it instead of creating our own encoding.

* It's a [well-established standard](https://en.wikipedia.org/wiki/UTF-8#Adoption):
  * The UTF-8 specification is far more mature than ours (it's used by ~97% of the web).
  * Programming languages such as [Lua](https://www.lua.org/manual/5.4/manual.html#6.5), [Python](https://docs.python.org/3/library/stdtypes.html#str) and [Rust](https://doc.rust-lang.org/stable/std/?search=utf8) have built-in UTF-8 libraries.
  * There is research on optimizing the decoding, validation and computation of UTF-8 strings.
* It has all the advantages of a [self-synchronizing code](https://en.wikipedia.org/wiki/Self-synchronizing_code):
  * The start of a code point can be found from a random position (by skipping at most 3 CBs).
  * More generally, it is a [prefix code](https://en.wikipedia.org/wiki/Prefix_code), since every byte has one of the following prefixes:
    * `0xxx xxxx` is a single-byte code point;
    * `10xx xxxx` is a continuation byte (CB);
    * `110x xxxx` is the leading byte to a 2-byte code point (followed by 1 CB);
    * `1110 xxxx` is the leading byte to a 3-byte code point (followed by 2 CBs), and;
    * `1111 0xxx` is the leading byte to a 4-byte code point (followed by 3 CBs).
* It also has some other very interesting properties:
  * Multiple bit flips in the prefix of a byte are detectable.
  * Multiple bit flips that create an [overlong encoding](https://en.wikipedia.org/wiki/UTF-8#Overlong_encodings) are also detectable.

**Reply by augustoteixeira** [reply]
I have considered this alternative but you bring an interesting point: perhaps having an implementation for all languages is an important benefit.

But the two other points that you mentioned don't sound like an advantage to me.

Our requirements are very different from those of UTF-8 (which are supposed to work well in faulty channels).
- Self-syncing: we are not decoding a sequence of code-points. We are decoding a single one. Actually, to make UTF-8 self-syncing they lose several bits.
- Resilience to bit-flips: our environment is deterministic, so bits shouldn't flip. But suppose that a bit does flip or that an input is malformed so that it corresponds to a faulty code. In this case, we want to raise an error and stop the processing of the input immediately. I am afraid that an UTF-8 library would try to be smart and silently interprets the input for us as something random. This could be dangerous.

The loss of bits mentioned in the second point is not serious, because this waste only appears for codepoints with more than one byte and UTF-8 already gives thousands of codepoints at 2 bytes. So this is fine.

The biggest concern I have is the second point. Perhaps we could do some tests and see what happens.

Indeed having implementations in several languages is a very positive aspect.

**Reply by guidanoli** [reply]
> Actually, to make UTF-8 self-syncing they lose several bits.

There's actually something called [Variable-Length Quantity](https://en.wikipedia.org/wiki/Variable-length_quantity), which encodes integers of arbitrary length and is self-syncing, but does not have a prefix in the leading byte to indicate the integers' length, so we can't jump to the leading byte of the next code point in constant time (which is something our "home-brewed" encoding and UTF-8 leverage). It is able to encode the same amount of code points per byte as our encoding (2^7 code points per byte). So, let's compare our options:

| Encoding | Biggest differential | Self-syncing | Length in leading byte | Lengths | Encoding bits per byte |
| -- | -- | -- | -- | -- | -- |
| VLQ | Arbitrary length | Yes | No | 1-∞ | 7 |
| Home-brewed | | No | Yes | 1-4 | 7 |
| UTF-8 | Maturity and adoption | Yes | Yes | 1-4 | 7, 5.5, ~5.3, 5.25 [1] |

[1] for 1-byte, 2-byte, 3-byte and 4-byte encodings, respectively

> The biggest concern I have is the second point.

Yes, we should run some experiments, but I believe UTF-8 is able to detect faulty characters more often than our encoding, because of the requirements of the leading byte of an n-byte code point being followed by n-1 continuation bytes.

As a side note, random strings get less likely to be valid UTF-8 strings the longer they are [2]:

| Length | Probability |
| -- | -- |
| 2 | 0.03591 |
| 4 | 0.027588 |
| 8 | 0.004001 |
| 15 | 0.000081 |

[2] Duerst, M. (1977). The Properties and Promises of UTF-8. In 11th Unicode Conference.

> Indeed having implementations in several languages is a very positive aspect.

Yes, and these implementations must have been peer-reviewed and tested thoroughly, so they should be less prone to bugs than our home-brewed solution.

**Reply by augustoteixeira** [reply]
I agree, let us just check that the implementations we use (mainly Rust) are not trying to be super smart.

Knowing how Rust API's are designed, I would say that they would report any errors in the parsing.

With this confirmed, we can move on and change the CIP!
